pkg/expectation/expectation.go

```diff
@@ -0,0 +1,98 @@
+package expectation
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pracucci/cortex-load-generator/pkg/gen"
+	"github.com/prometheus/common/model"
+)
+
+const maxComparisonDelta = 0.001
+
+type Validator func(start, end time.Time, result []model.SamplePair) error
+
+// Expectation describes which responses are expected from the TSDB, and starting from when.
+// The writer client can update the expectation and the reader can make assertions on it.
+type Expectation struct {
+	ValidFrom time.Time
+	Data      map[string]*Sequence // validations which specify, for a selector, all the specified samples
+	Funcs     map[string]Validator // validations which specify, for a selector, a function to run on the response
+	sync.Mutex
+}
+
+func NewExpectation() *Expectation {
+	return &Expectation{
+		Data:  make(map[string]*Sequence),
+		Funcs: make(map[string]Validator),
+	}
+}
+
+// Adjustment lets a write client adjust the expectation
+type Adjustment func(e *Expectation)
+
+func (e *Expectation) Adjust(fn Adjustment) {
+	e.Lock()
+	fn(e)
+	e.Unlock()
+}
+
+func (e *Expectation) Validate(selector string, start, end time.Time, result []model.SamplePair) error {
+	e.Lock()
+	defer e.Unlock()
+	now := time.Now()
+	if e.ValidFrom.After(now) {
+		time.Sleep(e.ValidFrom.Sub(now))
+	}
+	exp, ok := e.Data[selector]
+	if ok {
+		exp.TrimLeft(start.UnixMilli())
```

Comment:
Why do you need to do this?

Reply:
Oh I see now, there could be more samples in this expectation and you only want to validate those between `[start,end]`.

Reply:
yeah basically, it's the query client that knows start and end that it's querying for.
the write client can't "predict" those parameters so it doesn't try to prune old samples out of the expectation